# Table of Contents
### 1. Import Librairies
### 2. Import Data
### 3. Merge Data Frames
### 4. Export Data

# 1. Import Librairies

In [1]:
#Import libraries
import pandas as pd
import numpy as np
import os

# 2. Import Data

In [2]:
#Creating path
path = r'C:\Users\manev\Documents\Agentür für Arbeit\CareerFoundry\Data Immersion\Achievement 4\Instacart Basket Analysis'

In [3]:
#Create data frame for the "orders_checked" data set
df_ords_checked = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'), index_col = False)

In [4]:
#Create data frame for the "products_checked" data set
df_prods_checked = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index_col = False)

In [5]:
#Create data frame for the "orders_products_prior" data set
df_ords_prior = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders_products_prior.csv'), index_col = False)

In [6]:
#Checking the output
df_ords_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [6]:
df_ords_prior.shape

(32434489, 4)

In [7]:
#Checking df_ords_checked
df_ords_checked

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

There is a common key column between these two dataframes: the product_id.

In [8]:
#Checking the dimension of df_ords_checked
df_ords_checked.shape

(3421083, 6)

# 3. Merge Data Frames

In [10]:
#Merging df_ords_checked and df_ords_prior
df_merged_large = df_ords_checked.merge(df_ords_prior, on = 'order_id', indicator = True)

In [11]:
#Checking the merged dataframe
df_merged_large.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  
0                     NaN         196                  1          0   both  
1                     NaN       14084                  2          0   both  
2                     NaN       12427                  3          0   both  
3                     NaN       26088                  4          0   both  
4                     NaN       26405                  5          0   both

In [12]:
#Checking if it's a full match
df_merged_large['_merge'].value_counts()

_merge
both          32434489
left_only            0
right_only           0
Name: count, dtype: int64

In [13]:
#Merging df_ords_checked and df_ords_prior using an outer merge
df_merged_large_example = df_ords_checked.merge(df_ords_prior, on = 'order_id', how = 'outer', indicator = True)

In [14]:
#Checking if the dataframe createrd with an outer merge is a full match
df_merged_large_example['_merge'].value_counts()

_merge
both          32434489
left_only       206209
right_only           0
Name: count, dtype: int64

For the Instacart project, we only need the data from the df_ords_checked to be matched with the data from df_ords_prior, which is the case in df_merged_large. Df_merged_large has '32434489' values, just like df_ords_checked.
We will therefore be using df_merged_large from now on.

In [15]:
#Merging df_ords_checked and df_ords_prior without the merge flag
df_orders_products_combined = df_ords_checked.merge(df_ords_prior, on = 'order_id')

In [16]:
#Checking the shape of df_orders_products_combined
df_orders_products_combined.shape

(32434489, 9)

# 4. Export Data

In [17]:
# Export df_merged_large to pkl

df_orders_products_combined.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_combined.pkl'))